### PR TITLE
bulk request update with removing uid

### PR DESCRIPTION
**Pull request by bpathak-ons**
### What

* removing additional uid field while preparing bulk request
* refactoring for update action

### How to review

* check, by running unit test, bulk request is not having uid field
* extensive test may include running service locally and see if update action is performed for duplicate document
* CI Pass

### Who can review

* anyone

### CODE REVIEW

**Review by nshumoogum — DISMISSED** [review]

File: cmd/producer/main.go [hunk 1]
```diff
@@ -78,6 +78,11 @@ func main() {
 func scanEvent(scanner *bufio.Scanner) *models.SearchDataImportModel {
 	fmt.Println("--- [Send Kafka PublishedContent] ---")
 
+	fmt.Println("Type the uid")
```
Comment (justinpjose) [inline cmd/producer/main.go:81]:
```suggestion
	fmt.Println("Type the UID")
```

File: handler/handler.go [hunk 2]
```diff
@@ -54,94 +54,48 @@ func (batchHandler BatchHandler) sendToES(ctx context.Context, esDestURL string,
...
 	}
 
-	var bulkCreateRes models.EsBulkResponse
-	err = json.Unmarshal(jsonCreateResponse, &bulkCreateRes)
-	if err != nil {
-		log.Error(ctx, "error unmarshaling jsonCreateResponse", err)
-		return err
+	var bulkRes models.EsBulkResponse
+	if err := json.Unmarshal(jsonUpsertResponse, &bulkRes); err != nil {
+		log.Error(ctx, "error unmarshaling json", err)
```
Comment (justinpjose) [inline handler/handler.go:81]: Is there a reason why the error is not returned at this point?

File: handler/handler.go [hunk 3]
```diff
@@ -54,94 +54,48 @@ func (batchHandler BatchHandler) sendToES(ctx context.Context, esDestURL string,
 
 	log.Info(ctx, "bulk events into ES starts")
 	target := len(events)
-	documentList := make(map[string]models.SearchDataImportModel)
 
-	var bulkcreate []byte
+	var bulkupsert []byte
 	for _, event := range events {
-		if event.Title == "" {
```
Comment (justinpjose) [inline handler/handler.go:61]: Is there any reason why this validation is dismissed? 
And should there be any other validations in place?

File: handler/handler.go [hunk 4]
```diff
@@ -54,94 +54,48 @@ func (batchHandler BatchHandler) sendToES(ctx context.Context, esDestURL string,
...
-				updateBulkRequestBody, err := prepareEventForBulkRequestBody(ctx, &esEvent, "update")
-				if err != nil {
-					log.Error(ctx, "error in preparing the bulk for update", err)
-					continue
-				}
-				bulkupdate = append(bulkupdate, updateBulkRequestBody...)
-			} else if resCreateItem["create"].Status == 201 {
+	if bulkRes.Errors {
+		for _, resUpsertItem := range bulkRes.Items {
+			if resUpsertItem["create"].Status == 409 {
```
Comment (justinpjose) [inline handler/handler.go:85]: The value `"create"` should be a constant. Then if that value changes in the future, then it is a matter of just changing the value of constant from the one place

File: handler/handler_test.go [hunk 5]
```diff
@@ -121,15 +101,13 @@ func clientMock(doFunc func(ctx context.Context, request *http.Request) (*http.R
 	}
 }
 
-func TestHandleWithTwoEventsBothEventCreated(t *testing.T) {
+func TestHandleWithEventsCreated(t *testing.T) {
 
 	var count int
 	Convey("Given a handler configured with sucessful es updates for all two events is success", t, func() {
```
Comment (justinpjose) [inline handler/handler_test.go:107]:
```suggestion
	Convey("Given a handler configured with successful es updates for all two events is success", t, func() {
```

File: handler/handler_test.go [hunk 6]
```diff
@@ -121,15 +101,13 @@ func clientMock(doFunc func(ctx context.Context, request *http.Request) (*http.R
 	}
 }
 
-func TestHandleWithTwoEventsBothEventCreated(t *testing.T) {
+func TestHandleWithEventsCreated(t *testing.T) {
 
 	var count int
```
Comment (justinpjose) [inline handler/handler_test.go:106]: This is not required
```suggestion
	
```

File: handler/handler_test.go [hunk 7]
```diff
@@ -148,20 +126,14 @@ func TestHandleWithTwoEventsBothEventCreated(t *testing.T) {
 	})
```
Comment (justinpjose) [inline handler/handler_test.go:126]: On line 121, 
The Convey message should be 
`Then the error is nil`

The remainder `only create bulk is called but not update bulk request` is incorrect as only one call "upsert" is always made

Also on line 123, the test to check `count` is `1` does not inform what call is made. It only informs whether a call is made or not

File: handler/handler_test.go [hunk 8]
```diff
@@ -148,20 +126,14 @@ func TestHandleWithTwoEventsBothEventCreated(t *testing.T) {
...
 }
 
-func TestHandleWithTwoEventsWithOneEventCreateSuccessAndOtherUpdateSuccess(t *testing.T) {
+func TestHandleWithEventsUpdated(t *testing.T) {
 
 	var count int
 	Convey("Given a handler configured with sucessful es updates for two events with one create error", t, func() {
 
 		doFuncWithValidResponse := func(ctx context.Context, req *http.Request) (*http.Response, error) {
 			count++
```
Comment (justinpjose) [inline handler/handler_test.go:135]: Same as before

File: handler/handler_test.go (hunk 7 above)
Comment (justinpjose) [reply handler/handler_test.go:126]: Please note that count will be useful if you are testing that a call has been made to ES, irrespective of the type of call

**Review by justinpjose — CHANGES_REQUESTED** [review]
Some comments. Happy to do a call to go through this

File: handler/handler.go (hunk 3 above)
Comment (bpathak-ons) [reply handler/handler.go:61]: reintroduced using uid instead

File: handler/handler.go [hunk 9]
```diff
@@ -79,16 +87,17 @@ func (batchHandler BatchHandler) sendToES(ctx context.Context, esDestURL string,
 	var bulkRes models.EsBulkResponse
 	if err := json.Unmarshal(jsonUpsertResponse, &bulkRes); err != nil {
 		log.Error(ctx, "error unmarshaling json", err)
+		return err
 	}
 	if bulkRes.Errors {
 		for _, resUpsertItem := range bulkRes.Items {
-			if resUpsertItem["create"].Status == 409 {
+			if resUpsertItem[create].Status == 409 {
```
Comment (justinpjose) [inline handler/handler.go:94]: The variable name should be more explicit. Just `create` doesn't tell us what it is

File: handler/handler.go [hunk 10]
```diff
@@ -57,6 +60,11 @@ func (batchHandler BatchHandler) sendToES(ctx context.Context, esDestURL string,
 
 	var bulkupsert []byte
 	for _, event := range events {
+		if event.UID == "" {
+			log.Info(ctx, "No uid for inbound kafka event, no transformation possible")
```
Comment (justinpjose) [inline handler/handler.go:64]: the message should be all lower-case
`"no uid for inbound kafka event, no transformation possible"`

File: handler/handler.go [hunk 11]
```diff
@@ -57,6 +60,11 @@ func (batchHandler BatchHandler) sendToES(ctx context.Context, esDestURL string,
 
 	var bulkupsert []byte
 	for _, event := range events {
+		if event.UID == "" {
```
Comment (justinpjose) [inline handler/handler.go:63]: Is this the only validation available?

**Review by justinpjose — CHANGES_REQUESTED** [review]
Some more changes

File: handler/handler.go (hunk 11 above)
Comment (bpathak-ons) [reply handler/handler.go:63]: validations will be address in separate tasks, no current requirements

**Review by justinpjose — APPROVED** [review]

**Pull request merged**